File: src/main/kotlin/web/url/AbstractUrl.kt
package web.url

import tracking.website.Strategy
import java.net.URL

abstract class AbstractUrl(
    private val url: String,
    private val strategy: Strategy,
) : Url {
    private val netUrl = URL(url)

    override fun getUrl() = url
    override fun getStrategy() = strategy
    override fun getHost(): String = netUrl.host
}

File: symfony/src/Utils/Web/Url/AbstractUrl.php
<?php

declare(strict_types=1);

namespace App\Utils\Web\Url;

use App\Utils\Web\UrlStrategy\Strategies;
use App\Utils\Web\UrlStrategy\Strategy;
use Override;

abstract readonly class AbstractUrl implements Url
{
    public Strategy $strategy;

    public function __construct(
        public string $url,
        ?Strategy $strategy = null,
    ) {
        $this->strategy = $strategy ?? Strategies::getFor($this->url);
    }

    #[Override]
    public function getOriginalUrl(): string
    {
        return $this->url;
    }

    #[Override]
    public function getUrl(): string
    {
        return $this->url;
    }

    #[Override]
    public function getStrategy(): Strategy
    {
        return $this->strategy;
    }
}
